python/lsst/cp/pipe/cpLinearitySolve.py

```diff
@@ -1211,3 +1118,1241 @@ def debugFit(self, stepname, xVector, yVector, yModel, mask, ampName):
+            if groupingColumn not in ptc.auxValues:
+                raise ValueError(f"Config requests grouping by {groupingColumn}, "
+                                 "but this column is not available in ptc.auxValues.")
+
+            uGroupValues = np.unique(ptc.auxValues[groupingColumn])
+            for i, uGroupValue in enumerate(uGroupValues):
+                groupingValues[ptc.auxValues[groupingColumn] == uGroupValue] = i
+    else:
+        means = np.zeros((nPair, len(ptc.ampNames)))
+        exptimes = np.zeros_like(means)
+        for i, ampName in enumerate(ptc.ampNames):
+            means[:, i] = ptc.rawMeans[ampName] * ptc.gain[ampName]
+            exptimes[:, i] = ptc.rawExpTimes[ampName]
+        detMeans = np.nanmean(means, axis=1)
+        detExptimes = np.nanmean(exptimes, axis=1)
+
+        if autoGroupingUseExptime:
+            abscissa = detExptimes
+        else:
+            abscissa = ptc.photoCharges[ptc.ampNames[0]].copy()
+            # Set illegal photocharges to NaN.
+            abscissa[abscissa < minPhotodiodeCurrent] = np.nan
+
+        ratio = detMeans / abscissa
+        ratio /= np.nanmedian(ratio)
+
+        # Adjust those that are above threshold so they fall into the
+        # largest group.
+        above = (detMeans > autoGroupingMaxSignalFraction*np.nanmax(detMeans))
+        maxIndex = np.argmax(detMeans[~above])
+        ratio[above] = ratio[maxIndex]
+
+        # The clustering of ratios into groups is performed with a simple
+        # algorithm based on sorting and looking for the largest gaps.
+        # See https://stackoverflow.com/a/18385795
+        st = np.argsort(ratio)
+        stratio = ratio[st]
+        delta = stratio[1:] - stratio[0: -1]
+
+        transitions, = np.where(delta > autoGroupingThreshold)
+        if len(transitions) > 0:
+            ratioCuts = stratio[transitions] + autoGroupingThreshold/2.
+
+            for i in range(len(transitions) + 1):
+                if i == 0:
+                    inGroup, = np.where(ratio < ratioCuts[i])
+                elif i == len(transitions):
+                    inGroup, = np.where(ratio > ratioCuts[i - 1])
+                else:
+                    inGroup, = np.where((ratio > ratioCuts[i - 1]) & (ratio < ratioCuts[i]))
+                groupingValues[inGroup] = i
+
+        # Ensure out-of-range photoCharges/exptimes are in their own group.
+        # These are masked later on.
+        groupingValues[~np.isfinite(abscissa)] = -1
+        # And put the high ones in the max group.
+        groupingValues[above] = groupingValues[maxIndex]
+
+    return groupingValues
+
+
+def _computeTurnoffAndMax(
+    abscissa,
+    ordinate,
+    initialMask,
+    groupingValues,
+    ampName="UNKNOWN",
+    minSignalFitLinearityTurnoff=1000.0,
+    maxFracLinearityDeviation=0.01,
+    log=None,
+):
+    """Compute the turnoff and max signal.
+
+    Parameters
+    ----------
+    minSignalFitLinearityTurnoff : `float`
```

Comment:
I was going to ask if this could be moved until after the position based arguments, but I think it's a duplicate that can be deleted.